### PR TITLE
feat(router): Added JWT authentication

### PR DESCRIPTION
**Pull request by racnan**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

Added JWT authentication. Also added implementation for retrieving merchantAccount from JWT.

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
This will be used in future for authentication purposes.

## How did you test it?
Local testing with a route set to auth with JWT.

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/Cargo.toml [hunk 1]
```diff
@@ -67,6 +67,7 @@ tokio = { version = "1.24.1", features = ["macros", "rt-multi-thread"] }
 url = { version = "2.3.1", features = ["serde"] }
 utoipa = { git = "https://github.com/SanchithHegde/utoipa", branch = "retain-property-order" }
 uuid = { version = "1.2.2", features = ["serde", "v4"] }
+jsonwebtoken = "8.2.0"
```
Comment (Abhicodes-crypto) [inline crates/router/Cargo.toml:70]: Please arrange this in alphabetic order.

File: crates/router/src/services/authentication.rs [hunk 2]
```diff
@@ -101,6 +102,32 @@ impl AuthenticateAndFetch<storage::MerchantAccount> for PublishableKeyAuth {
...
+        state: &AppState,
+    ) -> RouterResult<storage::MerchantAccount> {
+        let mut token = get_jwt(request_headers)?;
+        token = strip_jwt_token(token)?;
+        let payload = decode_jwt::<JwtAuthPayloadFetchMerchantAccount>(token, state)?;
+        state
+            .store
+            .find_merchant_account_by_merchant_id(&payload.merchant_id)
+            .await
+            .map_err(|_| errors::ApiErrorResponse::InvalidJwtToken.into())
```
Comment (jarnura) [inline crates/router/src/services/authentication.rs:127]: Don't discard the error, convert the error into report and attach `InvalidJwtToken`.

File: crates/router/src/services/authentication.rs [hunk 3]
```diff
@@ -183,6 +223,21 @@ pub async fn is_ephemeral_auth(
...
+
+pub fn decode_jwt<T>(token: &str, state: &AppState) -> RouterResult<T>
+where
+    T: serde::de::DeserializeOwned,
+{
+    let secret = state.conf.keys.jwt_secret.as_bytes();
+    let key = DecodingKey::from_secret(secret);
+    decode::<T>(token, &key, &Validation::new(Algorithm::HS256))
+        .map(|decoded| decoded.claims)
+        .map_err(|_| errors::ApiErrorResponse::InvalidJwtToken.into())
```
Comment (jarnura) [inline crates/router/src/services/authentication.rs:238]: Don't discard the error, convert the error into report and attach InvalidJwtToken.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/services/authentication.rs [hunk 4]
```diff
@@ -101,6 +102,32 @@ impl AuthenticateAndFetch<storage::MerchantAccount> for PublishableKeyAuth {
...
+        &self,
+        request_headers: &HeaderMap,
+        state: &AppState,
+    ) -> RouterResult<storage::MerchantAccount> {
+        let mut token = get_jwt(request_headers)?;
+        token = strip_jwt_token(token)?;
+        let payload = decode_jwt::<JwtAuthPayloadFetchMerchantAccount>(token, state)?;
+        state
+            .store
+            .find_merchant_account_by_merchant_id(&payload.merchant_id)
```
Comment (Abhicodes-crypto) [inline crates/router/src/services/authentication.rs:125]: Just a question, if a DB call has to happen for authentication then what is the advantage of using JWT token ? Wouldn't  a similar auth like an api key suffice ?

**Review by jarnura — APPROVED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]
LGTM

File: crates/router/src/services/authentication.rs (hunk 4 above)
Comment (racnan) [reply crates/router/src/services/authentication.rs:125]: - JWT can contain more information in payload and not strictly be attached to getting merchant account. This is just one of the implementation. Whereas, API key in our case is strictly attached to a merchant account.
- API key is generally used for S2S calls, whereas JWT will be used for client to server call, where token will be issued for every session and can be expired (i guess API keys cant be expired based on time).

File: crates/router/src/services/authentication.rs [hunk 5]
```diff
@@ -183,6 +223,22 @@ pub async fn is_ephemeral_auth(
     Ok(Box::new(MerchantIdAuth(ephemeral_key.merchant_id)))
 }
 
+fn is_jwt_auth(headers: &HeaderMap) -> bool {
+    headers.get("Authorization").is_some()
```
Comment (SanchithHegde) [inline crates/router/src/services/authentication.rs:227]: Use `crate::headers::AUTHORIZATION` instead of this string literal.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

**Review by SanchithHegde — CHANGES_REQUESTED** [review]
My bad, should have requested changes.

**Review by SanchithHegde — APPROVED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]

**Comment by racnan** [comment]
@jarnura please merge this.

**Review by jarnura — APPROVED** [review]

**Pull request merged**